File: app/src/main/java/me/echeung/moemoekyun/service/RadioService.java
package me.echeung.moemoekyun.service;

import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.media.AudioManager;
import android.net.ConnectivityManager;
import android.os.Binder;
import android.os.Bundle;
import android.os.IBinder;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.RatingCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.util.Log;
import android.view.KeyEvent;
import android.widget.Toast;

import java.text.ParseException;
import java.util.Calendar;

import androidx.annotation.NonNull;
import me.echeung.moemoekyun.App;
import me.echeung.moemoekyun.BuildConfig;
import me.echeung.moemoekyun.R;
import me.echeung.moemoekyun.client.api.callback.FavoriteSongCallback;
import me.echeung.moemoekyun.client.api.library.Jpop;
import me.echeung.moemoekyun.client.api.library.Kpop;
import me.echeung.moemoekyun.client.model.Song;
import me.echeung.moemoekyun.client.socket.Socket;
import me.echeung.moemoekyun.client.socket.response.UpdateResponse;
import me.echeung.moemoekyun.client.stream.Stream;
import me.echeung.moemoekyun.ui.activity.MainActivity;
import me.echeung.moemoekyun.util.AlbumArtUtil;
import me.echeung.moemoekyun.util.PreferenceUtil;
import me.echeung.moemoekyun.util.SongActionsUtil;
import me.echeung.moemoekyun.util.system.TimeUtil;
import me.echeung.moemoekyun.viewmodel.RadioViewModel;

public class RadioService extends Service implements Socket.Listener, AlbumArtUtil.Callback, SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String TAG = RadioService.class.getSimpleName();

    private static final String APP_PACKAGE_NAME = BuildConfig.APPLICATION_ID;

    private static final int MILLISECONDS_IN_SECOND = 1000;

    public static final String PLAY_PAUSE = APP_PACKAGE_NAME + ".play_pause";
    public static final String STOP = APP_PACKAGE_NAME + ".stop";
    public static final String TOGGLE_FAVORITE = APP_PACKAGE_NAME + ".toggle_favorite";
    public static final String LIBRARY_JPOP = APP_PACKAGE_NAME + ".library_jpop";
    public static final String LIBRARY_KPOP = APP_PACKAGE_NAME + ".library_kpop";
    public static final String UPDATE = APP_PACKAGE_NAME + ".update";
    public static final String TIMER_STOP = APP_PACKAGE_NAME + ".timer_stop";

    private static final long MEDIA_SESSION_ACTIONS = PlaybackStateCompat.ACTION_PLAY
            | PlaybackStateCompat.ACTION_PAUSE
            | PlaybackStateCompat.ACTION_PLAY_PAUSE
            | PlaybackStateCompat.ACTION_STOP
            | PlaybackStateCompat.ACTION_SET_RATING;

    private final IBinder binder = new ServiceBinder();

    private AppNotification notification;
    private Stream stream;
    private Socket socket;

    private volatile MediaSessionCompat mediaSession;
    private final Object mediaSessionLock = new Object();

    private AudioManager audioManager;
    private AudioManager.OnAudioFocusChangeListener audioFocusChangeListener;
    private boolean wasPlayingBeforeLoss;

    private BroadcastReceiver intentReceiver;
    private boolean receiverRegistered = false;

    private boolean isFirstConnectivityChange = true;

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public void onCreate() {
        AlbumArtUtil.registerListener(this);

        initBroadcastReceiver();
        initMediaSession();
        initAudioManager();

        stream = App.getRadioClient().getStream();
        socket = App.getRadioClient().getSocket();

        stream.setListener(new Stream.Listener() {
            @Override
            public void onStreamPlay() {
                App.getRadioViewModel().setIsPlaying(true);

                updateNotification();
                updateMediaSessionPlaybackState();
            }

            @Override
            public void onStreamPause() {
                App.getRadioViewModel().setIsPlaying(false);

                updateNotification();
                updateMediaSessionPlaybackState();
            }

            @Override
            public void onStreamStop() {
                audioManager.abandonAudioFocus(audioFocusChangeListener);

                stopForeground(true);
                stopSelf();

                App.getPreferenceUtil().clearSleepTimer();
                App.getRadioViewModel().setIsPlaying(false);

                updateMediaSessionPlaybackState();
            }
        });

        socket.setListener(this);
        socket.connect();

        App.getPreferenceUtil().registerListener(this);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startID) {
        handleIntent(intent);

        return START_NOT_STICKY;
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        if (!isPlaying()) {
            stopSelf();
        }
    }

    @Override
    public void onDestroy() {
        AlbumArtUtil.unregisterListener(this);

        stop();
        socket.disconnect();
        stream.removeListener();

        if (receiverRegistered) {
            unregisterReceiver(intentReceiver);
            receiverRegistered = false;
        }

        destroyMediaSession();

        App.getPreferenceUtil().unregisterListener(this);

        super.onDestroy();
    }

    public boolean isStreamStarted() {
        return stream.isStarted();
    }

    public boolean isPlaying() {
        return stream.isPlaying();
    }

    @Override
    public void onSocketReceive(UpdateResponse.Details info) {
        RadioViewModel viewModel = App.getRadioViewModel();

        viewModel.setListeners(info.getListeners());
        viewModel.setRequester(info.getRequester());
        viewModel.setEvent(info.getEvent());

        if (info.getQueue() != null) {
            viewModel.setQueueSize(info.getQueue().getInQueue());
            viewModel.setInQueueByUser(info.getQueue().getInQueueByUser());
            viewModel.setQueuePosition(info.getQueue().getInQueueBeforeUser());
        }

        Calendar startTime = null;
        try {
            startTime = TimeUtil.toCalendar(info.getStartTime());
        } catch (ParseException e) {
            e.printStackTrace();
        }

        viewModel.setCurrentSong(info.getSong(), startTime);
        viewModel.setLastSong(info.getLastPlayed().get(0));
        viewModel.setSecondLastSong(info.getLastPlayed().get(1));

        updateMediaSession();
        updateNotification();
    }

    @Override
    public void onSocketFailure() {
        App.getRadioViewModel().reset();
        updateNotification();
    }

    public MediaSessionCompat getMediaSession() {
        return mediaSession;
    }

    private void updateMediaSession() {
        Song currentSong = App.getRadioViewModel().getCurrentSong();

        if (currentSong == null) {
            mediaSession.setMetadata(null);
            return;
        }

        MediaMetadataCompat.Builder metaData = new MediaMetadataCompat.Builder()
                .putString(MediaMetadataCompat.METADATA_KEY_TITLE, currentSong.getTitleString())
                .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, currentSong.getArtistsString())
                .putString(MediaMetadataCompat.METADATA_KEY_ALBUM, currentSong.getAlbumsString())
                .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, currentSong.getDuration() * MILLISECONDS_IN_SECOND);

        if (App.getPreferenceUtil().shouldShowLockscreenAlbumArt()) {
            Bitmap albumArt = AlbumArtUtil.getCurrentAlbumArt();
            if (albumArt != null && !AlbumArtUtil.isDefaultAlbumArt()) {
                metaData.putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, albumArt);
                updateNotification();
            }
        }

        synchronized (mediaSessionLock) {
            mediaSession.setMetadata(metaData.build());
            updateMediaSessionPlaybackState();
        }
    }

    private void updateMediaSessionPlaybackState() {
        // Play/pause state
        PlaybackStateCompat.Builder stateBuilder = new PlaybackStateCompat.Builder()
                .setActions(MEDIA_SESSION_ACTIONS)
                .setState(isStreamStarted()
                                ? isPlaying() ? PlaybackStateCompat.STATE_PLAYING : PlaybackStateCompat.STATE_PAUSED
                                : PlaybackStateCompat.STATE_STOPPED,
                        App.getRadioViewModel().getCurrentSongProgress(), 1);

        // Favorite action
        if (App.getAuthUtil().isAuthenticated()) {
            Song currentSong = App.getRadioViewModel().getCurrentSong();
            int favoriteIcon = currentSong == null || !currentSong.isFavorite() ?
                    R.drawable.ic_star_border_white_24dp :
                    R.drawable.ic_star_white_24dp;

            stateBuilder.addCustomAction(new PlaybackStateCompat.CustomAction.Builder(
                    TOGGLE_FAVORITE, getString(R.string.favorite), favoriteIcon)
                    .build());
        }

        if (mediaSession.isActive()) {
            mediaSession.setPlaybackState(stateBuilder.build());
        }
    }

    @Override
    public void onAlbumArtReady(Bitmap bitmap) {
        updateMediaSession();
    }

    private void updateNotification() {
        if (isStreamStarted()) {
            if (notification == null) {
                notification = new AppNotification(this);
            }

            notification.update();
        } else {
            stopForeground(true);
        }
    }

    private boolean handleIntent(Intent intent) {
        if (intent == null) return true;

        String action = intent.getAction();
        if (action != null) {
            switch (action) {
                case RadioService.PLAY_PAUSE:
                    togglePlayPause();
                    break;

                case RadioService.STOP:
                    stop();
                    break;

                case RadioService.TOGGLE_FAVORITE:
                    favoriteCurrentSong();
                    break;

                case RadioService.UPDATE:
                case SongActionsUtil.REQUEST_EVENT:
                    socket.update();
                    break;

                case RadioService.TIMER_STOP:
                    timerStop();
                    break;

                // Pause when headphones unplugged
                case AudioManager.ACTION_AUDIO_BECOMING_NOISY:
                    if (App.getPreferenceUtil().shouldPauseOnNoisy()) {
                        pause();
                    }
                    break;

                // Headphone media button action
                case Intent.ACTION_MEDIA_BUTTON:
                    Bundle extras = intent.getExtras();
                    if (extras == null) {
                        return false;
                    }

                    KeyEvent keyEvent = (KeyEvent) extras.get(Intent.EXTRA_KEY_EVENT);
                    if (keyEvent == null || keyEvent.getAction() != KeyEvent.ACTION_DOWN) {
                        return false;
                    }

                    switch (keyEvent.getKeyCode()) {
                        case KeyEvent.KEYCODE_HEADSETHOOK:
                        case KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE:
                            togglePlayPause();
                            break;
                        case KeyEvent.KEYCODE_MEDIA_PLAY:
                            play();
                            break;
                        case KeyEvent.KEYCODE_MEDIA_PAUSE:
                            pause();
                            break;
                        case KeyEvent.KEYCODE_MEDIA_STOP:
                            stop();
                            break;
                        case KeyEvent.KEYCODE_MEDIA_NEXT:
                        case KeyEvent.KEYCODE_MEDIA_PREVIOUS:
                            // Do nothing
                            break;
                    }
                    break;

                case MainActivity.AUTH_EVENT:
                    socket.reconnect();
                    if (!App.getAuthUtil().isAuthenticated()) {
                        App.getRadioViewModel().setIsFavorited(false);
                        updateNotification();
                    }
                    break;

                case ConnectivityManager.CONNECTIVITY_ACTION:
                    // Ignore the initial sticky broadcast on app start
                    if (isFirstConnectivityChange) {
                        isFirstConnectivityChange = false;
                        break;
                    }

                    socket.reconnect();
            }
        }

        updateNotification();
        return true;
    }

    private void initMediaSession() {
        synchronized (mediaSessionLock) {
            mediaSession = new MediaSessionCompat(this, APP_PACKAGE_NAME, null, null);
            mediaSession.setRatingType(RatingCompat.RATING_HEART);
            mediaSession.setCallback(new MediaSessionCompat.Callback() {
                @Override
                public void onPlay() {
                    play();
                }

                @Override
                public void onPause() {
                    pause();
                }

                @Override
                public void onStop() {
                    stop();
                }

                @Override
                public void onSkipToNext() {
                }

                @Override
                public void onSkipToPrevious() {
                }

                @Override
                public void onSeekTo(long pos) {
                }

                @Override
                public void onSetRating(RatingCompat rating) {
                    favoriteCurrentSong();
                }

                @Override
                public boolean onMediaButtonEvent(Intent mediaButtonEvent) {
                    return handleIntent(mediaButtonEvent);
                }

                @Override
                public void onCustomAction(@NonNull String action, Bundle extras) {
                    switch (action) {
                        case TOGGLE_FAVORITE:
                            favoriteCurrentSong();
                            updateMediaSessionPlaybackState();
                            break;

                        default:
                            Log.d(TAG, "Unsupported action: " + action);
                            break;
                    }
                }

                @Override
                public void onPlayFromSearch(String query, Bundle extras) {
                    // We don't support searching for specific things since it's just a radio stream
                    // so just toggle playback
                    togglePlayPause();
                }

                @Override
                public void onPlayFromMediaId(String mediaId, Bundle extras) {
                    super.onPlayFromMediaId(mediaId, extras);

                    // Handles changing library mode via Android Auto
                    switch (mediaId) {
                        case LIBRARY_JPOP:
                            App.getRadioClient().changeLibrary(Jpop.NAME);
                            break;

                        case LIBRARY_KPOP:
                            App.getRadioClient().changeLibrary(Kpop.NAME);
                            break;
                    }
                }
            });

            mediaSession.setFlags(MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS
                    | MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS);

            mediaSession.setActive(true);
        }
    }

    private void destroyMediaSession() {
        synchronized (mediaSessionLock) {
            if (mediaSession != null) {
                mediaSession.setActive(false);
                mediaSession.release();
            }
        }
    }

    private void initBroadcastReceiver() {
        intentReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                handleIntent(intent);
            }
        };

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(RadioService.PLAY_PAUSE);
        intentFilter.addAction(RadioService.STOP);
        intentFilter.addAction(RadioService.TOGGLE_FAVORITE);
        intentFilter.addAction(RadioService.UPDATE);
        intentFilter.addAction(SongActionsUtil.REQUEST_EVENT);
        intentFilter.addAction(AudioManager.ACTION_AUDIO_BECOMING_NOISY);
        intentFilter.addAction(Intent.ACTION_MEDIA_BUTTON);
        intentFilter.addAction(MainActivity.AUTH_EVENT);
        intentFilter.addAction(ConnectivityManager.CONNECTIVITY_ACTION);

        registerReceiver(intentReceiver, intentFilter);
        receiverRegistered = true;
    }

    private void initAudioManager() {
        audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        audioFocusChangeListener = focusChange -> {
            switch (focusChange) {
                case AudioManager.AUDIOFOCUS_GAIN:
                    stream.unduck();
                    if (wasPlayingBeforeLoss) {
                        play();
                    }
                    break;

                case AudioManager.AUDIOFOCUS_LOSS:
                case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                    wasPlayingBeforeLoss = isPlaying();
                    if (wasPlayingBeforeLoss &&
                            (App.getPreferenceUtil().shouldPauseAudioOnLoss()
                                    || AutoMediaBrowserService.isCarUiMode(this))) {
                        pause();
                    }
                    break;

                case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK:
                    wasPlayingBeforeLoss = isPlaying();
                    if (App.getPreferenceUtil().shouldDuckAudio()) {
                        stream.duck();
                    }
                    break;
            }
        };
    }

    private void togglePlayPause() {
        if (isPlaying()) {
            pause();
        } else {
            play();
        }
    }

    private void play() {
        // Request audio focus for playback
        int result = audioManager.requestAudioFocus(audioFocusChangeListener,
                AudioManager.STREAM_MUSIC,
                AudioManager.AUDIOFOCUS_GAIN);

        if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            stream.play();
        }
    }

    private void pause() {
        stream.pause();
    }

    private void stop() {
        stream.stop();
    }

    private void timerStop() {
        stream.fadeOut();
    }

    private void favoriteCurrentSong() {
        Song currentSong = App.getRadioViewModel().getCurrentSong();
        if (currentSong == null) return;

        int songId = currentSong.getId();
        if (songId == -1) return;

        if (!App.getAuthUtil().isAuthenticated()) {
            showLoginRequiredToast();
            return;
        }

        boolean isCurrentlyFavorite = currentSong.isFavorite();

        FavoriteSongCallback callback = new FavoriteSongCallback() {
            @Override
            public void onSuccess() {
                Song currentSong = App.getRadioViewModel().getCurrentSong();
                if (currentSong.getId() == songId) {
                    App.getRadioViewModel().setIsFavorited(!isCurrentlyFavorite);
                }

                Intent favIntent = new Intent(SongActionsUtil.FAVORITE_EVENT);
                sendBroadcast(favIntent);

                updateNotification();
                updateMediaSessionPlaybackState();
            }

            @Override
            public void onFailure(String message) {
                Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT).show();
            }
        };

        App.getRadioClient().getApi().toggleFavorite(String.valueOf(songId), isCurrentlyFavorite, callback);
    }

    private void showLoginRequiredToast() {
        Toast.makeText(getApplicationContext(), R.string.login_required, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        switch (key) {
            case PreferenceUtil.PREF_LOCKSCREEN_ALBUMART:
            case PreferenceUtil.PREF_GENERAL_ROMAJI:
                updateMediaSession();
                break;
        }
    }

    public class ServiceBinder extends Binder {
        public RadioService getService() {
            return RadioService.this;
        }
    }

}

File: app/src/main/kotlin/me/echeung/moemoekyun/service/RadioService.kt
package me.echeung.moemoekyun.service

import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.media.AudioManager
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Binder
import android.os.Bundle
import android.os.IBinder
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.RatingCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.view.KeyEvent
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.merge
import kotlinx.coroutines.flow.onEach
import logcat.LogPriority
import logcat.asLog
import logcat.logcat
import me.echeung.moemoekyun.BuildConfig
import me.echeung.moemoekyun.R
import me.echeung.moemoekyun.client.RadioClient
import me.echeung.moemoekyun.client.api.Library
import me.echeung.moemoekyun.client.api.socket.Socket
import me.echeung.moemoekyun.client.api.socket.response.UpdateResponse
import me.echeung.moemoekyun.client.api.auth.AuthUtil
import me.echeung.moemoekyun.client.stream.Stream
import me.echeung.moemoekyun.service.notification.MusicNotifier
import me.echeung.moemoekyun.ui.activity.auth.AuthActivityUtil
import me.echeung.moemoekyun.util.AlbumArtUtil
import me.echeung.moemoekyun.util.PreferenceUtil
import me.echeung.moemoekyun.util.SongActionsUtil
import me.echeung.moemoekyun.util.ext.connectivityManager
import me.echeung.moemoekyun.util.ext.launchIO
import me.echeung.moemoekyun.util.ext.launchUI
import me.echeung.moemoekyun.util.ext.toast
import me.echeung.moemoekyun.util.system.TimeUtil
import me.echeung.moemoekyun.viewmodel.RadioViewModel
import org.koin.android.ext.android.inject
import java.text.ParseException
import java.util.Calendar

class RadioService : Service() {

    private val scope = MainScope()

    private val radioClient: RadioClient by inject()
    private val stream: Stream by inject()
    private val socket: Socket by inject()

    private val albumArtUtil: AlbumArtUtil by inject()
    private val authUtil: AuthUtil by inject()
    private val preferenceUtil: PreferenceUtil by inject()

    private val radioViewModel: RadioViewModel by inject()

    private val binder = ServiceBinder()

    private var notifier = MusicNotifier(this, albumArtUtil)

    @Volatile
    var mediaSession: MediaSessionCompat? = null
        private set
    private val mediaSessionLock = Any()

    private var intentReceiver: BroadcastReceiver? = null
    private var receiverRegistered = false

    val isStreamStarted: Boolean
        get() = stream.isStarted

    val isPlaying: Boolean
        get() = stream.isPlaying

    val isLoading: Boolean
        get() = stream.isLoading

    override fun onBind(intent: Intent): IBinder? = binder

    override fun onCreate() {
        initBroadcastReceiver()
        initNetworkStateCallback()
        initMediaSession()

        merge(preferenceUtil.shouldPreferRomaji().asFlow(), preferenceUtil.shouldShowLockscreenAlbumArt().asFlow())
            .onEach { updateMediaSession() }
            .launchIn(scope)

        albumArtUtil.flow
            .onEach { updateMediaSession() }
            .launchIn(scope)

        stream.state
            .onEach {
                when (it) {
                    Stream.State.Loading -> {
                        radioViewModel.isLoading = true
                        updateNotification()
                    }
                    Stream.State.Playing -> {
                        radioViewModel.isPlaying = true
                        updateNotification()
                    }
                    Stream.State.Paused -> {
                        radioViewModel.isPlaying = false
                        updateNotification()
                    }
                    Stream.State.Stopped -> {
                        stopForeground(true)
                        stopSelf()

                        preferenceUtil.sleepTimer().delete()
                        radioViewModel.isPlaying = false
                    }
                }

                updateMediaSessionPlaybackState()
            }
            .launchIn(scope)

        socket.state
            .onEach {
                when (it) {
                    is Socket.State.Update -> it.info?.let { data -> onSocketReceive(data) }
                    is Socket.State.Error -> onSocketFailure()
                }
            }
            .launchIn(scope)

        socket.connect()
    }

    override fun onStartCommand(intent: Intent, flags: Int, startID: Int): Int {
        handleIntent(intent)

        return START_NOT_STICKY
    }

    override fun onTaskRemoved(rootIntent: Intent) {
        if (!isPlaying) {
            stopSelf()
        }
    }

    override fun onDestroy() {
        stream.stop()

        socket.disconnect()

        if (receiverRegistered) {
            unregisterReceiver(intentReceiver)
            receiverRegistered = false
        }

        destroyMediaSession()

        super.onDestroy()
    }

    private fun onSocketReceive(info: UpdateResponse.Details?) {
        radioViewModel.listeners = info!!.listeners
        radioViewModel.setRequester(info.requester)
        radioViewModel.event = info.event

        var startTime: Calendar? = null
        try {
            startTime = TimeUtil.toCalendar(info.startTime!!)
        } catch (e: ParseException) {
            logcat(LogPriority.ERROR) { "Error parsing time: ${e.asLog()}" }
        }

        // Check if current song is favorited
        if (info.song != null && authUtil.isAuthenticated) {
            launchIO {
                val favoritedSongIds = radioClient.api.isFavorite(listOf(info.song.id))
                if (info.song.id in favoritedSongIds && radioViewModel.currentSong?.id == info.song.id) {
                    radioViewModel.isFavorited = true
                }
            }
        }

        radioViewModel.setCurrentSong(info.song, startTime)
        radioViewModel.lastSong = info.lastPlayed!![0]
        radioViewModel.secondLastSong = info.lastPlayed[1]

        albumArtUtil.updateAlbumArt(info.song)

        updateMediaSession()
        updateNotification()
    }

    private fun onSocketFailure() {
        radioViewModel.reset()
        updateNotification()
    }

    private fun updateMediaSession() {
        val currentSong = radioViewModel.currentSong

        if (currentSong == null) {
            mediaSession!!.setMetadata(null)
            return
        }

        val metaData = MediaMetadataCompat.Builder()
            .putString(MediaMetadataCompat.METADATA_KEY_TITLE, currentSong.getTitleString())
            .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, currentSong.getArtistsString())
            .putString(MediaMetadataCompat.METADATA_KEY_ALBUM, currentSong.getAlbumsString())
            .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, (currentSong.duration * MILLISECONDS_IN_SECOND).toLong())

        if (preferenceUtil.shouldShowLockscreenAlbumArt().get()) {
            val albumArt = albumArtUtil.currentAlbumArt
            if (albumArt != null && !albumArtUtil.isDefaultAlbumArt) {
                metaData.putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, albumArt)
                updateNotification()
            }
        }

        synchronized(mediaSessionLock) {
            mediaSession!!.setMetadata(metaData.build())
            updateMediaSessionPlaybackState()
        }
    }

    private fun updateMediaSessionPlaybackState() {
        // Play/pause state
        val stateBuilder = PlaybackStateCompat.Builder()
            .setActions(MEDIA_SESSION_ACTIONS)
            .setState(
                when {
                    !isStreamStarted -> PlaybackStateCompat.STATE_STOPPED
                    isLoading -> PlaybackStateCompat.STATE_CONNECTING
                    isPlaying -> PlaybackStateCompat.STATE_PLAYING
                    else -> PlaybackStateCompat.STATE_PAUSED
                },
                radioViewModel.currentSongProgress,
                1f,
            )

        // Favorite action
        if (authUtil.isAuthenticated) {
            val currentSong = radioViewModel.currentSong
            val favoriteIcon = if (currentSong == null || !currentSong.favorite) {
                R.drawable.ic_star_border_24dp
            } else {
                R.drawable.ic_star_24dp
            }

            stateBuilder.addCustomAction(
                PlaybackStateCompat.CustomAction.Builder(
                    TOGGLE_FAVORITE,
                    getString(R.string.favorite),
                    favoriteIcon
                )
                    .build()
            )
        }

        if (mediaSession!!.isActive) {
            mediaSession!!.setPlaybackState(stateBuilder.build())
        }
    }

    private fun updateNotification() {
        if (isStreamStarted) {
            notifier.update(radioViewModel.currentSong, authUtil.isAuthenticated)
        } else {
            stopForeground(true)
        }
    }

    private fun handleIntent(intent: Intent?): Boolean {
        if (intent == null) return true

        when (intent.action) {
            PLAY_PAUSE -> stream.toggle()

            STOP -> stream.stop()

            TOGGLE_FAVORITE -> favoriteCurrentSong()

            UPDATE, SongActionsUtil.REQUEST_EVENT -> socket.update()

            TIMER_STOP -> stream.fadeOut()

            // Pause when headphones unplugged
            AudioManager.ACTION_AUDIO_BECOMING_NOISY -> if (preferenceUtil.shouldPauseOnNoisy()) {
                stream.pause()
            }

            // Headphone media button action
            Intent.ACTION_MEDIA_BUTTON -> {
                val extras = intent.extras ?: return false

                val keyEvent = extras.get(Intent.EXTRA_KEY_EVENT) as KeyEvent?
                if (keyEvent == null || keyEvent.action != KeyEvent.ACTION_DOWN) {
                    return false
                }

                when (keyEvent.keyCode) {
                    KeyEvent.KEYCODE_HEADSETHOOK, KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE -> stream.toggle()
                    KeyEvent.KEYCODE_MEDIA_PLAY -> stream.play()
                    KeyEvent.KEYCODE_MEDIA_PAUSE -> stream.pause()
                    KeyEvent.KEYCODE_MEDIA_STOP -> stream.stop()
                }
            }

            AuthActivityUtil.AUTH_EVENT -> {
                socket.reconnect()
                if (!authUtil.isAuthenticated) {
                    radioViewModel.isFavorited = false
                    updateNotification()
                }
            }
        }

        updateNotification()
        return true
    }

    private fun initMediaSession() {
        synchronized(mediaSessionLock) {
            mediaSession = MediaSessionCompat(this, APP_PACKAGE_NAME, null, null).apply {
                setRatingType(RatingCompat.RATING_HEART)
                setCallback(object : MediaSessionCompat.Callback() {
                    override fun onPlay() {
                        stream.play()
                    }

                    override fun onPause() {
                        stream.pause()
                    }

                    override fun onStop() {
                        stream.stop()
                    }

                    override fun onSkipToNext() {}

                    override fun onSkipToPrevious() {}

                    override fun onSeekTo(pos: Long) {}

                    override fun onSetRating(rating: RatingCompat?) {
                        favoriteCurrentSong()
                    }

                    override fun onMediaButtonEvent(mediaButtonEvent: Intent): Boolean {
                        return handleIntent(mediaButtonEvent)
                    }

                    override fun onCustomAction(action: String?, extras: Bundle?) {
                        when (action) {
                            TOGGLE_FAVORITE -> {
                                favoriteCurrentSong()
                                updateMediaSessionPlaybackState()
                            }

                            else -> logcat { "Unsupported action: $action" }
                        }
                    }

                    override fun onPlayFromSearch(query: String?, extras: Bundle?) {
                        if (!query.isNullOrEmpty()) {
                            when (query.lowercase()) {
                                "jpop", "j-pop" -> onPlayFromMediaId(LIBRARY_JPOP, extras)
                                "kpop", "k-pop" -> onPlayFromMediaId(LIBRARY_KPOP, extras)
                            }
                        }

                        if (!isPlaying) {
                            stream.play()
                        }
                    }

                    override fun onPlayFromMediaId(mediaId: String?, extras: Bundle?) {
                        super.onPlayFromMediaId(mediaId, extras)

                        // Handles changing library mode via Android Auto
                        when (mediaId) {
                            LIBRARY_JPOP -> radioClient.changeLibrary(Library.jpop)
                            LIBRARY_KPOP -> radioClient.changeLibrary(Library.kpop)
                        }
                    }
                })
                isActive = true
            }
        }
    }

    private fun destroyMediaSession() {
        synchronized(mediaSessionLock) {
            mediaSession?.isActive = false
            mediaSession?.release()
        }
    }

    private fun initBroadcastReceiver() {
        intentReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                handleIntent(intent)
            }
        }

        val intentFilter = IntentFilter().apply {
            addAction(PLAY_PAUSE)
            addAction(STOP)
            addAction(TOGGLE_FAVORITE)
            addAction(UPDATE)
            addAction(SongActionsUtil.REQUEST_EVENT)
            addAction(AudioManager.ACTION_AUDIO_BECOMING_NOISY)
            addAction(Intent.ACTION_MEDIA_BUTTON)
            addAction(AuthActivityUtil.AUTH_EVENT)
        }

        registerReceiver(intentReceiver, intentFilter)
        receiverRegistered = true
    }

    private fun initNetworkStateCallback() {
        val builder = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
            .addTransportType(NetworkCapabilities.TRANSPORT_VPN)

        val callback: ConnectivityManager.NetworkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                super.onAvailable(network)
                socket.reconnect()
            }

            override fun onLost(network: Network) {
                super.onLost(network)
                socket.disconnect()
            }
        }

        connectivityManager.registerNetworkCallback(builder.build(), callback)
    }

    private fun favoriteCurrentSong() {
        val song = radioViewModel.currentSong ?: return

        val songId = song.id
        if (songId == -1) return

        if (!authUtil.isAuthenticated) {
            toast(R.string.login_required)
            return
        }

        val isCurrentlyFavorite = song.favorite

        launchIO {
            try {
                radioClient.api.toggleFavorite(songId)

                val currentSong = radioViewModel.currentSong
                if (currentSong!!.id == songId) {
                    radioViewModel.isFavorited = !isCurrentlyFavorite
                }
                song.favorite = !isCurrentlyFavorite

                launchUI {
                    val favIntent = Intent(SongActionsUtil.FAVORITE_EVENT).apply {
                        setPackage(packageName)
                    }
                    sendBroadcast(favIntent)

                    updateNotification()
                    updateMediaSessionPlaybackState()
                }
            } catch (e: Exception) {
                launchUI { toast(e.message) }
            }
        }
    }

    inner class ServiceBinder : Binder() {
        val service: RadioService
            get() = this@RadioService
    }

    companion object {
        const val PLAY_PAUSE = "$APP_PACKAGE_NAME.play_pause"
        const val STOP = "$APP_PACKAGE_NAME.stop"
        const val TOGGLE_FAVORITE = "$APP_PACKAGE_NAME.toggle_favorite"
        const val LIBRARY_JPOP = "$APP_PACKAGE_NAME.library_jpop"
        const val LIBRARY_KPOP = "$APP_PACKAGE_NAME.library_kpop"
        const val UPDATE = "$APP_PACKAGE_NAME.update"
        const val TIMER_STOP = "$APP_PACKAGE_NAME.timer_stop"
    }
}

private const val APP_PACKAGE_NAME = BuildConfig.APPLICATION_ID
private const val MILLISECONDS_IN_SECOND = 1000

private const val MEDIA_SESSION_ACTIONS = (
    PlaybackStateCompat.ACTION_PLAY
        or PlaybackStateCompat.ACTION_PAUSE
        or PlaybackStateCompat.ACTION_PLAY_PAUSE
        or PlaybackStateCompat.ACTION_STOP
        or PlaybackStateCompat.ACTION_SET_RATING
    )
